### PR TITLE
Fixes an issue which can result in gaps in the Reader

### LINKED ISSUE #5491: Reader: missing posts
### Expected behavior

All posts that I see at https://wordpress.com/ with a browser also appear in Reader on WPAndroid.

### Actual behavior

Posts are missing in Reader on WPAndroid compared to viewing on the web.

### Steps to reproduce the behavior

1. Go to Reader in WPAndroid and browse posts.
1. Look for any large gaps in time such as from 5h to 2d.
1. Compare to https://wordpress.com/ in a browser to see if any posts are missing while viewing Reader in WPAndroid.

Note: I don't check Reader on WPAndroid every day. I can't remember exactly, but it's likely I haven't checked Reader on WPAndroid in 2d and that could be why the gap starts at that time for me.

screenshot_20170323-161509

screen shot 2017-03-23 at thu mar 23 4 16 59 pm

Maybe related: #3908

### PR DESCRIPTION
**Pull request by oguzkocer**
Fixes #5491. This PR fixes an issue which can sometimes result in a gap in the Reader. Note that while investigating #5491 this is one of the first issues I found, but I wouldn't be surprised if there is another issue that can result in a gap as I didn't audit the whole feature.

When we fetch the first page of the reader feed, we compare it to the existing posts in the DB and if at least one matches, we assume we fetched all the posts and there is no gap. The problem is that a post might be fetched as part of a different list and we'll act as if there is no gap even though there might be. The fix is fairly straightforward, we just check if the post exists for the given reader tag.

Please note that this PR will not fix the existing gaps in the Reader. These gaps will auto-resolve themselves fairly quickly, so I think resetting the table to fix the gaps is necessary.

---
**To test**
I think it's worth reproducing this issue first in `develop` and trying this branch again to check if the fix worked. Here are the steps to reproduce the issue:
1. Go to reader and refresh the list
2. Follow some random sites from another client, but make sure they’ll have at least 20 posts (in total) published before the first post in your list
3. Switch to the discover list
4. Follow at least one site you weren’t following before (the more you follow the bigger the gap will be)
5. Switch back to following list
6. It’ll auto refresh, but if it doesn’t, pull to refresh
7. Scroll 19 posts
8. Notice that there is a gap between the first post you got in `Step 1` and the 20th post in your list

---

The easiest way to test this would be to either create a new account or update the existing account and remove all the sites you're following and only follow `oguzkocertest.wordpress.com` whose last post is from a while back, so once you go into `Discover` and start following different sites, it'll be very easy to ensure at least 20 posts will be on top of the last post and check the gap.

Here is a gif of me reproducing this issue: https://cloudup.com/crG9Caq-GLH and here is the gif of the same steps after the fix: https://cloudup.com/cV8mgczYEBC

---
Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java [hunk 1]
```diff
@@ -398,11 +406,11 @@ public static boolean postExists(long blogId, long postId) {
...
 
     /*
-     * returns true if any posts in the passed list exist in this list
+     * returns true if any posts in the passed list exist in this list for the given tag
      */
-    public static boolean hasOverlap(ReaderPostList posts) {
+    public static boolean hasOverlap(ReaderPostList posts, ReaderTag tag) {
         for (ReaderPost post : posts) {
-            if (postExists(post.blogId, post.postId)) {
+            if (ReaderPostTable.postExistsForReaderTag(post.blogId, post.postId, tag)) {
```
Comment (theck13) [inline WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java]:
```suggestion
            if (postExistsForReaderTag(post.blogId, post.postId, tag)) {
```

We can remove the `ReaderPostTable.` prefix for this method call.  If you prefer it for clarity, just let me know.

File: RELEASE-NOTES.txt [hunk 2]
```diff
@@ -9,6 +9,7 @@
 * The Site → Blog Posts → View WebView will no longer show any Calypso UI
 * Fixed an issue with Twitter publicize connections
 * Fixed an issue where the site icon wasn't updating correctly after switching to another site
+* Fixes an issue which can result in creating gaps in Reader
```
Comment (theck13) [inline RELEASE-NOTES.txt]: It wasn't entirely clear what "gaps" meant when I read this.  I first thought it was the margin/padding of the layout.  Users might not know what "gaps" refers to either, but if we intend to change this after review from Editorial, we can leave it as-is for now.

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java (hunk 1 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java]: I personally prefer having the prefix for static methods, but I don't mind it either way, so if you'd like we can make the change.

File: RELEASE-NOTES.txt (hunk 2 above)
Comment (oguzkocer) [reply RELEASE-NOTES.txt]: That makes sense, it wouldn't hurt to improve it even before it goes to the Editorial. How about something like this:

```
* Fixes an issue which can result in missing posts in the Reader
```

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java (hunk 1 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java]: If that's your preference, we can leave it.  It's not a problem at all.

File: RELEASE-NOTES.txt (hunk 2 above)
Comment (theck13) [reply RELEASE-NOTES.txt]: That works for me.  Users and Editorial will understand that better.

Comment (oguzkocer) [reply RELEASE-NOTES.txt]: Fixed in 200064e4332f7c237bd1cfec7a0df6ec592c0a65.

File: WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java (hunk 1 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java]: It looks like everywhere else we are using the methods without the prefix, so let's make the change to keep things consistent. Thanks for bringing this up.

**Pull request merged**